pilot/pkg/leaderelection/leaderelection.go

```diff
@@ -142,8 +153,13 @@ func (l *LeaderElection) create() (*k8sleaderelection.LeaderElector, error) {
 	if l.prioritized {
 		// Function to use to decide whether this revision should steal the existing lock.
 		config.KeyComparison = func(currentLeaderRevision string) bool {
```

Comment:
This function is pretty complex and hard to understand and extremely critical. There is only a few inputs to it and its a pure, simple function. Can we split this out and unit test ~all the cases.

It would also be great to keep a list of all past `KeyComparison` functions, and run them against the current one, to ensure we do not have any inputs for which two different pods will both attempt to steal; if we are in this case its a critical bug. Would make me re-assured to have a test case to verify that is not the case.

Reply:
Not sure I understand what you mean by `keep a list of all past KeyComparison functions`. Can you clarify?

I will add more unit tests to try to make sure that all combinations of remote/local, default/non-default, default/default and non-default/non-default are tested.

Reply:
What I mean is that in a given cluster, we will have Istiods running with:
* No prioritized election
* The current HEAD  KeyComparison function
* This new KeyComparison fn
* Any future KeyComparison functions

If ANY of these different versions both think they should steal a leader lock, we will end up with an infinite loop of stealing the lock and probably break the cluster.

Reply:
I've improved the algorithm and added testing of all the combinations. PTAL.